File: langchain_py/app/1/quickstart.py
from app.config import settings
from dataclasses import dataclass
from typing import Literal
from langchain.agents import create_agent
from langchain.tools import tool, ToolRuntime
from langchain_openai import ChatOpenAI
from langchain.agents.structured_output import ToolStrategy
from langgraph.checkpoint.memory import InMemorySaver


'''
标准的agent创建和使用流程： 
1.定义提示词
2.定义工具
3.构建chat model
4.结构化输出
5.memory
6.创建agent
'''

# 1.定义提示词
SYSTEM_PROMPT = """你是天气助手,如果被问到天气问题，请先确定地点，然后调用相关工具获取实际天气"""


# 2.定义工具

@dataclass
class Context:
    """Custom runtime context schema."""
    user_id: str

@tool
def get_weather_for_location(city: str) -> str:
    """获取指定城市的天气"""
    return f"{city}总是晴日"

@tool
def get_user_location(runtime: ToolRuntime[Context]) -> str:
    """根据用户ID获取用户位置"""
    user_id = runtime.context.user_id
    return "北京" if user_id == "1" else "上海"

# 3.构建chat model
# 符合openai规范的api,可以使用langchain_openai。我们使用 SiliconFlow 提供的 GLM-4.7 模型
# 注意：尽量使用一些新模型，一些旧模型可能会存在一些特性不支持
model = ChatOpenAI(
    model=settings.glm_model,
    base_url=settings.siliconflow_base_url,
    api_key=settings.siliconflow_api_key,
    temperature=0.9,
    max_tokens=5000,
    timeout=60,
)

# 4.结构化输出
# dataclass 和 Pydantic 都是支持的，用来定义结构化输出的格式。
@dataclass
class ResponseFormat:
    """agent的响应格式"""
    # 一语双关的回答 (必要)
    punny_response: str
    # 和天气相关的信息点（可选）
    weather_conditions: str | None = None
    # 字符串，用于描述响应的长度，取值为"short"、"medium"、"long"之一
    length: Literal["short", "medium", "long"] = "short"

# 5.memory
checkpointer = InMemorySaver()


# 6.创建agent
agent = create_agent(
    model,
    system_prompt=SYSTEM_PROMPT,
    tools=[get_user_location, get_weather_for_location],
    context_schema=Context,
    response_format=ToolStrategy(ResponseFormat),
    checkpointer=checkpointer,
    # debug=True, # 开启debug模式，会打印出agent的运行过程
)

# `thread_id`一次会话的唯一标识符
config = {"configurable": {"thread_id": "1"}}

response = agent.invoke(
    {"messages": [{"role": "user", "content": "今天天气如何？"}]},
    config=config,
    context=Context(user_id="1")
)

print(response['structured_response'])
# ResponseFormat(punny_response='今天的天气晴朗温暖，阳光明媚，绝对是一个出去走走的好日子！☀️', weather_conditions='晴日', length='medium')

# 注意：我们可以用同一个`thread_id`继续这个对话.
response = agent.invoke(
    {"messages": [{"role": "user", "content": "thank you!"}]},
    config=config,
    context=Context(user_id="1")
)

print(response['structured_response'])
# ResponseFormat(punny_response='不用客气！很高兴能帮到你！希望你今天有个晴朗愉快的一天！☀️😊', weather_conditions=None, length='medium')

File: langchain_ts/app/1/quickstart.ts
import { settings } from "@/config";
import { MemorySaver } from "@langchain/langgraph";
import { ChatOpenAI } from "@langchain/openai";
import type { RunnableConfig } from "@langchain/core/runnables";
import { createAgent, tool, toolStrategy } from "langchain";
import { z } from "zod";

/*
标准的agent创建和使用流程：
1.定义提示词
2.定义工具
3.构建chat model
4.结构化输出
5.memory
6.创建agent
*/

// 1.定义提示词
const SYSTEM_PROMPT =
  "你是天气助手,如果被问到天气问题，请先确定地点，然后调用相关工具获取实际天气";

// 2.定义工具
const get_weather_for_location = tool(({ city }: { city: string }) => {
  return `${city}总是晴日`;
}, {
  name: "get_weather_for_location",
  description: "获取指定城市的天气",
  schema: z.object({
    city: z.string(),
  }),
});

const get_user_location = tool(
  async (_: Record<string, never>, config?: RunnableConfig) => {
    const user_id = String(
      (config?.configurable as Record<string, unknown> | undefined)?.user_id ??
        "",
    );
    return user_id === "1" ? "北京" : "上海";
  },
  {
    name: "get_user_location",
    description: "根据用户ID获取用户位置",
    schema: z.object({}),
  },
);

// 3.构建chat model
// 符合openai规范的api,可以使用 @langchain/openai。我们使用 SiliconFlow 提供的 GLM-4.7 模型
// 注意：尽量使用一些新模型，一些旧模型可能会存在一些特性不支持
const model = new ChatOpenAI({
  model: settings.glm_model,
  apiKey: settings.siliconflow_api_key,
  configuration: {
    baseURL: settings.siliconflow_base_url,
  },
  temperature: 0.9,
  maxTokens: 5000,
  timeout: 60000,
});

// 4.结构化输出
// dataclass 和 Pydantic 都是支持的，用来定义结构化输出的格式。
const ResponseFormat = z.object({
  // 一语双关的回答 (必要)
  punny_response: z.string().describe("一语双关的回答（必要）"),
  // 和天气相关的信息点（可选）
  weather_conditions: z
    .string()
    .nullable()
    .optional()
    .describe("和天气相关的信息点（可选）"),
  // 字符串，用于描述响应的长度，取值为"short"、"medium"、"long"之一
  length: z
    .enum(["short", "medium", "long"])
    .default("short")
    .describe('响应长度，取值为"short"、"medium"、"long"之一'),
}).describe("agent的响应格式");

// 5.memory
const checkpointer = new MemorySaver();

// 6.创建agent
const agent = createAgent({
  model,
  systemPrompt: SYSTEM_PROMPT,
  tools: [get_user_location, get_weather_for_location],
  responseFormat: toolStrategy(ResponseFormat),
  checkpointer,
  // debug: true, # 开启debug模式，会打印出agent的运行过程
});

async function main() {
  // `thread_id`一次会话的唯一标识符
  const config = { configurable: { thread_id: "1", user_id: "1" } };

  const response1 = await agent.invoke(
    { messages: [{ role: "user", content: "今天天气如何？" }] },
    config,
  );
  console.log(response1.structuredResponse);
//   {
//     punny_response: "北京总是晴日，今天也要保持阳光好心情！",
//     weather_conditions: "晴朗",
//     length: "short",
//   }

  // 注意：我们可以用同一个`thread_id`继续这个对话.
  const response2 = await agent.invoke(
    { messages: [{ role: "user", content: "thank you!" }] },
    config,
  );
  console.log(response2.structuredResponse);
// {
//   punny_response: "不客气！祝你有个阳光灿烂的一天！",
//   length: "short",
// }
}

if (import.meta.main) {
  await main();
}
